src/hooks/useSound.ts

```diff
@@ -0,0 +1,370 @@
+ */
+function useSound({
+  url,
+  loop = false,
+  defaultVolume = 1,
+  defaultPlaybackRate = 1,
+  onLoad,
+  onPlay,
+  onPause,
+  onEnd,
+}: useSoundProps): UseSoundReturnType {
+  const [volume] = useState(() =>
+    setValidRange(0, 1, defaultVolume, 'defaultVolume')
+  );
+  const [playbackRate] = useState(() =>
+    setValidRange(0.5, 4, defaultPlaybackRate, 'defaultPlaybackRate')
+  );
+  const [state, dispatch] = useReducer(reducer, {
+    ...initialState,
+    isLoop: loop || false,
+    volume,
+    playbackRate,
+  });
+
+  const audioRef = useRef<HTMLAudioElement | null>(null);
+
+  const safeAudioHelper = (action: SafeAudioAction) => {
+    if (audioRef.current) {
+      action(audioRef.current);
+    }
+  };
+
+  const cleanUpAudioRef = () => {
+    if (audioRef.current) {
+      audioRef.current.pause();
+      audioRef.current = null;
+    }
+  };
+
+  useEffect(() => {
+    audioRef.current = new Audio(url);
+
+    return () => {
+      cleanUpAudioRef();
+    };
+  }, [url]);
+
+  useEffect(() => {
+    if (!audioRef.current) {
+      return;
+    }
+
+    const cleanup = setupEventHandlers(
+      audioRef.current,
+      state.isLoop,
+      dispatch,
+      onLoad,
+      onPlay,
+      onPause,
+      onEnd
+    );
+
+    return () => {
+      cleanup();
+      cleanUpAudioRef();
+    };
+  }, [onLoad, onPlay, onPause, onEnd, state.isLoop]);
+
+  useEffect(() => {
+    safeAudioHelper((audioRef) => {
+      setupAudioElement(audioRef, state);
+    });
+    // eslint-disable-next-line react-hooks/exhaustive-deps
+  }, [state.volume, state.playbackRate, state.isLoop, state.isMuted]);
+
+  const play = () =>
+    safeAudioHelper((audioRef) => {
+      dispatch({ type: ACTION_TYPES.SET_IS_PLAYING, payload: true });
+      dispatch({ type: ACTION_TYPES.SET_IS_PAUSED, payload: false });
+      audioRef.play();
+    });
+
+  const pause = () =>
+    safeAudioHelper((audioRef) => {
+      dispatch({ type: ACTION_TYPES.SET_IS_PAUSED, payload: true });
+      audioRef.pause();
+    });
+
+  const stop = () =>
+    safeAudioHelper((audioRef) => {
+      dispatch({ type: ACTION_TYPES.SET_IS_PLAYING, payload: false });
+      dispatch({ type: ACTION_TYPES.SET_IS_PAUSED, payload: false });
+      dispatch({ type: ACTION_TYPES.SET_CURRENT_TIME, payload: 0 });
+      audioRef.pause();
+    });
+
+  const setPosition = (time: number) => {
+    safeAudioHelper((audioRef) => {
+      dispatch({ type: ACTION_TYPES.SET_CURRENT_TIME, payload: time });
+      audioRef.currentTime = time;
+    });
+  };
+
+  const setMute = () => {
+    dispatch({ type: ACTION_TYPES.SET_IS_MUTED, payload: true });
+  };
+
+  const setUnmute = () => {
+    dispatch({ type: ACTION_TYPES.SET_IS_MUTED, payload: false });
+  };
+
+  return {
+    ...state,
+    play,
+    pause,
+    stop,
+    setMute,
+    setUnmute,
+    setPosition,
+    setIsLoop: (isLoop) =>
+      dispatch({ type: ACTION_TYPES.SET_IS_LOOP, payload: isLoop }),
```

Comment:
```suggestion
    setIsLoop(isLoop) {
      dispatch({ type: ACTION_TYPES.SET_IS_LOOP, payload: isLoop })
    }
```
코딩컨벤션 Objects 메서드의 단축 구문!

Reply:
오옹 감사합니다 [a06d001](https://github.com/frontend-opensource-project/use-react-hooks/pull/37/commits/a06d0012e118a0aa928c54f2afbd7483a36546ae) 여기서 수정했어요!